// @base44/vite-plugin/src/processors/collection-item-id-processor.ts
import type { NodePath } from "@babel/traverse";
import type * as t from "@babel/types";
import { DATA_COLLECTION_ITEM_ID } from "../consts.js";
import {
  JSXAttributeUtils,
  ExpressionAnalysisUtils,
} from "./utils/shared-utils.js";

export class DataItemIdProcessor {
  private attributeUtils: JSXAttributeUtils;
  private expressionUtils: ExpressionAnalysisUtils;

  constructor(private types: typeof t) {
    this.attributeUtils = new JSXAttributeUtils(types);
    this.expressionUtils = new ExpressionAnalysisUtils(types);
  }

  process(path: NodePath<t.JSXOpeningElement>): void {
    if (this.attributeUtils.hasAttribute(path, DATA_COLLECTION_ITEM_ID)) {
      return;
    }

    const keyAttr = this.findKeyAttribute(path);
    if (!keyAttr) return;

    const expression = this.extractKeyExpression(keyAttr);
    if (!expression) return;

    if (!this.expressionUtils.isIdAccess(expression)) return;

    const optionalExpr = this.types.isMemberExpression(expression)
      ? this.expressionUtils.createOptionalChainExpression(expression)
      : expression;

    this.attributeUtils.addExpressionAttribute(
      path,
      DATA_COLLECTION_ITEM_ID,
      optionalExpr
    );
  }

  private findKeyAttribute(
    path: NodePath<t.JSXOpeningElement>
  ): t.JSXAttribute | null {
    for (const attr of path.node.attributes) {
      if (
        this.types.isJSXAttribute(attr) &&
        this.types.isJSXIdentifier(attr.name) &&
        attr.name.name === "key"
      ) {
        return attr;
      }
    }
    return null;
  }

  private extractKeyExpression(
    attr: t.JSXAttribute
  ): t.Expression | null {
    if (!attr.value) return null;

    if (this.types.isJSXExpressionContainer(attr.value)) {
      const expr = attr.value.expression;
      if (this.types.isExpression(expr)) return expr;
    }

    return null;
  }
}
